src/man/man3/sam_unload.3
.\" $Revision: 1.17 $
.ds ]W Sun Microsystems
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.TH sam_unload 3 "07 Jan 2009"
.SH NAME
sam_unload \- Unloads media on the removable media device
.SH SYNOPSIS
.LP
.BI "cc [ " "flag"
.BI " ... ] " "file"
.BI " ... -L/opt/SUNWsamfs/lib -lsamapi [ " "library" " ... ]"
.LP
.nf
.ft 3
#include "/opt/SUNWsamfs/include/samapi.h"
.ft
.fi
.LP
.BI "int sam_unload(ushort_t " "eq_number" ,
.BI "int " "wait_response" );
.SH DESCRIPTION
.PP
.B sam_unload(\|)
requests that the media be unloaded from
the device with equipment number
.I eq_number.
The device must be a removable media device or a robotic media changer.  The
device cannot be under the control of another process.
.PP
If the equipment number
.I eq_number
is a removable media device controlled by a robotic media changer, the medium
will be moved into storage.
.PP
If the equipment number
.I eq_number
is a robotic media changer, the unload
moves catalog entries from the robotic media changer's catalog to the
Historian's catalog.
.PP
The call will return immediately after issuing the command if
zero is specified for
.I wait_response
value.  Other values for
.I wait_response
will give undefined results.
.SH "RETURN VALUES"
Upon succesful completion a value of 0 is returned.
Otherwise, a value of \-1 is returned and
\f4errno\fP
is set to indicate the error.
.SH ERRORS
.PP
.B sam_unload(\|)
fails if one or more of the following are true:
.TP 20
.SB ER_DEVICE_NOT_READY
The specified
.I eq_number
device is not ready.
.TP
.SB ER_DEVICE_NOT_UNAVAILABLE
The specified
.I eq_number
device must be in the unavailable state (see \fBset_state\fR(8)).
.TP
.SB ER_DEVICE_USE_BY_ANOTHER
The specified
.I eq_number
device is busy and is being used by another process.
.TP
.SB ER_FIFO_PATH_LENGTH
The path and filename for the FIFO pipe is too long.
.TP
.SB ER_NO_DEVICE_FOUND
The device with equipment number
.I eq_number
is not available in this configuration.
.TP
.SB ER_NO_EQUIP_ORDINAL
.I eq_number
is not a defined equipment number in this configuration.
.TP
.SB ER_NO_MASTER_SHM
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master shared memory segment defined.
Check that the Sun StorEdge \%SAM-FS and Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_NO_MASTER_SHM_ATT
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master shared memory segment found.
Check that the Sun StorEdge \%SAM-FS and Sun \%SAM-QFS file systems are mounted.
.TP
.SB ER_NO_RESPONSE_FIFO
Unable to create the response FIFO pipe.
.TP
.SB ER_NOT_REMOV_MEDIA_DEVICE
The specified
.I eq_number
device is not a removable media device.
.TP
.SB ER_OPERATOR_NOT_PRIV
Operator does not have permission to unload removable media.
.SH FILES
.TP 20
.SB mcf
The configuration file for Sun StorEdge \%SAM-FS and Sun \%SAM-QFS file systems.
.SH SEE ALSO
.BR load (8),
.BR sam-robots (8),
.BR set_state (8),
.BR unload (8).
.PP
.BR sam_load (3).
.PP
.BR mcf (5).


src/lib/samapi/sam_unload.c
/*
 * sam_unload.c - Contains API for sam_unload() function
 *
 *
 */


/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.19 $"


#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <stropts.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/mtio.h>
#include <sys/stat.h>

#include "sam/types.h"
#include "sam/param.h"
#include "aml/device.h"
#include "aml/fifo.h"
#include "aml/proto.h"
#include "aml/sam_utils.h"
#include "aml/samapi.h"
#include "pub/sam_errno.h"



/*
 * sam_unload() - API function to unload
 *
 * Input parameters --
 *	eq_ord		Equipment ordinal
 *	wait_response	Wait for command response if nonzero
 *			-1	Wait forever
 *			> zero	Wait this many seconds for response
 *
 * Output parameters --
 *	errno		Set to error number if error encountered
 *
 * Return value --
 *	0		If successful in unloading the media
 *	-1		If error encountered,
 *			'errno' set to error number
 *
 */

int
sam_unload(
	ushort_t eq_ord,	/* Equipment ordinal */
	int wait_response)	/* Set nonzero to wait for command response */
{
	char	*fifo_path;		/* Path to FIFO pipe, */
					/* from shared memory */
	int	forever = FALSE;	/* Set TRUE if wait_response is -1 */
	int	wait = wait_response;	/* Duration of wait defined by caller */

	dev_ent_t	*device;
	operator_t	operator;	/* Data on operator */
	sam_cmd_fifo_t	cmd_block;


	/*
	 * If wait for response equals -1,
	 * set flag to indicate wait forever
	 */

	if (-1 == wait_response) {
		forever = TRUE;
	}

	/*
	 * Get device entry and FIFO path from master shared memory segment
	 */

	if (sam_get_dev(eq_ord, &device, &fifo_path, &operator) < 0) {
		return (-1);
	}

	if (!(IS_OPTICAL(device) || IS_TAPE(device) || IS_ROBOT(device))) {
		errno = ER_NOT_REMOV_MEDIA_DEVICE;
		return (-1);
	}

	/*
	 * If robot device and operator not privileged, return with error
	 */

	if (IS_ROBOT(device) && !SAM_OPER_SLOT(operator)) {
		errno = ER_OPERATOR_NOT_PRIV;
		return (-1);
	}

	/*
	 * If tape device is not available for SAM-FS,
	 * check device availability
	 */

	if (device->state >= DEV_UNAVAIL && IS_TAPE(device)) {
		int open_fd;

		if (((open_fd = open(device->name, O_RDONLY)) < 0) &&
		    (errno == EBUSY)) {
			if (forever || (wait > 0)) {
				while (forever || (wait-- > 0)) {
				/* N.B. Bad indentation to meet cstyle */
				/* requirements */
				if (((open_fd =
				    open(device->name, O_RDONLY)) < 0) &&
				    (errno == EBUSY)) {
					sleep(1);
				}
				else
				{
					break;
				}
				}
			}
			else
			{
				errno = ER_DEVICE_USE_BY_ANOTHER;
				return (-1);
			}
		}

		if (open_fd >= 0) {
			struct  mtop  tape_op;

			tape_op.mt_op = MTOFFL;
			tape_op.mt_count = 0;
			(void) ioctl(open_fd, MTIOCTOP, &tape_op);
			close(open_fd);
		}
		else
		{
			errno = ER_NO_DEVICE_FOUND;
			return (-1);
		}
	}

	if (!device->status.b.ready) {
		errno = ER_DEVICE_NOT_READY;
		return (-1);
	}

	/*
	 * Ready and send command message for unload of the device
	 */

	memset(&cmd_block, 0, sizeof (cmd_block));
	cmd_block.magic = CMD_FIFO_MAGIC;
	cmd_block.slot = ROBOT_NO_SLOT;
	cmd_block.cmd = CMD_FIFO_UNLOAD;
	cmd_block.eq = eq_ord;

	if (sam_send_cmd(&cmd_block, wait_response, fifo_path) < 0) {
		return (-1);
	}

	return (0);
}
